### PR TITLE
Let FastlanePty detect when externally invoked programs crash, harden it when using popen, and expose process statuses.

### PR DESCRIPTION
**Pull request by lacostej**
While looking into #21536, I noticed that tests seem to be missing on FastlanePty. Adding some

### Checklist

- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I see several green `ci/circleci` builds in the "All checks have passed" section of my PR ([connect CircleCI to GitHub](https://support.circleci.com/hc/en-us/articles/360008097173-Why-aren-t-pull-requests-triggering-jobs-on-my-organization-) if not)
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.
- [x] I've added or updated relevant unit tests.

Here's the test that is failing https://github.com/fastlane/fastlane/pull/21618/commits/e31a1d13ee0325cfb4a324b8ff6502934073cefe

### CODE REVIEW

File: fastlane_core/lib/fastlane_core/fastlane_pty.rb [hunk 1]
```diff
@@ -10,48 +10,78 @@ def exit_status
 
 module FastlaneCore
   class FastlanePtyError < StandardError
-    attr_reader :exit_status
-    def initialize(e, exit_status)
+    attr_reader :exit_status, :process_status
```
Comment (lacostej) [inline fastlane_core/lib/fastlane_core/fastlane_pty.rb:13]: I decided to expose the status, as it holds more information about the crash when it happens (e.g. the signal number, the pid). It's not fully necessary.

File: fastlane_core/lib/fastlane_core/fastlane_pty.rb [hunk 2]
```diff
@@ -10,48 +10,78 @@ def exit_status
...
+        raise FastlanePtyError.new(e, status.exitstatus || e.exit_status, status)
       end
-    rescue StandardError => e
-      # Wrapping any error in FastlanePtyError to allow
-      # callers to see and use $?.exitstatus that
-      # would usually get returned
-      raise FastlanePtyError.new(e, $?.exitstatus)
+    end
+
+    # to ease mocking
```
Comment (lacostej) [inline fastlane_core/lib/fastlane_core/fastlane_pty.rb:82]: didn't really find a way to mock the read-only `$?`

File: fastlane_core/spec/fastlane_pty_spec.rb [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        allow(status).to receive(:exitstatus) { nil }
+        allow(status).to receive(:signaled?) { true }
+
+        expect(FastlaneCore::FastlanePty).to receive(:require).with("pty").and_return(nil)
+        allow(FastlaneCore::FastlanePty).to receive(:process_status).and_return(status)
+
+        expect {
+          exit_status = FastlaneCore::FastlanePty.spawn("a path of a crasher exec") do |command_stdout, command_stdin, pid|
+          end
+          UI.message(exit_status)
```
Comment (AliSoftware) [inline fastlane_core/spec/fastlane_pty_spec.rb:56]: Leftover of debugging?

**Review by AliSoftware — APPROVED** [review]

File: fastlane_core/lib/fastlane_core/fastlane_pty.rb [hunk 4]
```diff
@@ -10,48 +10,78 @@ def exit_status
...
-          rescue Errno::ECHILD, PTY::ChildExited
-            # The process might have exited.
+            yield(command_stdout, command_stdin, pid)
+          rescue Errno::EIO
+            # Exception ignored intentionally.
+            # https://stackoverflow.com/questions/10238298/ruby-on-linux-pty-goes-away-without-eof-raises-errnoeio
+            # This is expected on some linux systems, that indicates that the subcommand finished
+            # and we kept trying to read, ignore it
+          ensure
+            begin
```
[the remark this answers is not in the document]
Comment (lacostej) [reply fastlane_core/lib/fastlane_core/fastlane_pty.rb:43]: looks like I can't get rid of this one

```
SyntaxError:
  /Users/lacostej/Code/OSS/fastlane/fastlane_core/lib/fastlane_core/fastlane_pty.rb:41: syntax error, unexpected `rescue', expecting `end'
          rescue Errno::ECHILD, PTY::ChildExi...
          ^~~~~~
```

File: fastlane_core/spec/fastlane_pty_spec.rb (hunk 3 above)
Comment (lacostej) [reply fastlane_core/spec/fastlane_pty_spec.rb]: oops. Fixed

**Pull request merged**